src/rust/engine/fs/store/src/local.rs

```diff
@@ -286,58 +302,94 @@ impl ByteStore {
     items: Vec<(Option<Digest>, Bytes)>,
     initial_lease: bool,
   ) -> Result<Vec<Digest>, String> {
+    let mut small_items = vec![];
+    let mut result = vec![];
+    for (digest, bytes) in items {
+      if matches!(entry_type, EntryType::File) && ByteStore::uses_large_file_store(bytes.len()) {
+        result.push(self.store_large_bytes(bytes, digest).await?);
+      } else {
+        small_items.push((digest, bytes));
+      }
+    }
+
     let dbs = match entry_type {
       EntryType::Directory => self.inner.directory_dbs.clone(),
       EntryType::File => self.inner.file_dbs.clone(),
     };
     // NB: False positive: we do actually need to create the Vec here, since `items` will move
     // before we use `lens`.
     #[allow(clippy::needless_collect)]
-    let lens = items
+    let lens = small_items
       .iter()
       .map(|(_, bytes)| bytes.len())
       .collect::<Vec<_>>();
     let fingerprints = dbs?
       .store_bytes_batch(
-        items
+        small_items
           .into_iter()
           .map(|(d, bytes)| (d.map(|d| d.hash), bytes))
           .collect(),
         initial_lease,
       )
       .await?;
 
-    Ok(
+    result.extend(
       fingerprints
         .into_iter()
         .zip(lens.into_iter())
-        .map(|(f, len)| Digest::new(f, len))
-        .collect(),
-    )
+        .map(|(f, len)| Digest::new(f, len)),
+    );
+
+    Ok(result)
+  }
+
+  /// Returns whether this file digest should/does use the "large file store" instead of the LMDB.
+  pub fn uses_large_file_store(len: usize) -> bool {
+    len > LARGE_FILE_SIZE_LIMIT
+  }
+
+  /// Assuming uses_large_file_store(digest) is true, return the path inside the store.
+  pub fn large_file_path(&self, digest: Digest) -> PathBuf {
+    assert!(ByteStore::uses_large_file_store(digest.size_bytes));
```

Comment:
Given the assert, this should probably not be a public function... anything exposed outside of the module should probably have a `Result` return type to account for accidental misuse.

Reply:
OK, let me see if I can make it internally public (crate public). It's used by the materialization path to choose when to symlink.